### ISSUE TITLE
Function 'MambaSplitConv1dScanCombinedFnBackward' returned nan values in its 0th output.

### ISSUE THREAD

**Issue by Liamhh20230518** [state-spaces/mamba#445]
<img width="1342" alt="Screenshot 2024-07-05 at 6 24 41 PM" src="https://github.com/state-spaces/mamba/assets/133937376/7b65968f-eca6-43ba-9ea1-1f0c6a869606">

mamba-ssm: 2.0.4

The loss is not nan, but the MambaSplitConv1dScanCombinedFnBackward' returned nan values in its 0th output

Thanks

**Comment by tridao** [comment]
Can you post a script that help us reproduce the error? E.g. save the tensors that produce the Nan?

**Comment by ArlenCHEN** [comment]
@tridao I encountered the same issue in my project. The loss is not nan, but it seems that the grad from one mamba func is nan.
<img width="1325" alt="Screenshot 2024-07-11 at 12 13 36" src="https://github.com/state-spaces/mamba/assets/25460810/cfe2507f-9715-4478-b820-d58201a7df60">

I checked the output of mamba_split_conv1d_scan_combined. It shows that the forward result is ok.
<img width="551" alt="Screenshot 2024-07-11 at 12 21 51" src="https://github.com/state-spaces/mamba/assets/25460810/fc34b787-e46f-4bd1-9713-361cb4548bce">

Please help here. Thanks!

**Comment by tridao** [comment]
Can you post a script that help us reproduce the error?

**Comment by liguochun0304** [comment]
> Can you post a script that help us reproduce the error?您能否发布一个脚本来帮助我们重现错误？

Hi, I noticed that you're encountering a similar issue to the one I'm facing. The problem I'm dealing with is that when I use the mamba module independently, it doesn't output any NaN values. However, when I integrate it into my model, NaN values start to appear. Interestingly, it works fine on one server, but when I switch to another server, the NaN values show up. Both machines are Dockerized, so I don't expect any environmental issues, and I've also double-checked things like nvcc and related configurations.

I’ve been trying to resolve this issue for a while now, but I haven’t been able to find a solution yet. It seems like this problem hasn’t been addressed by the author either. I was wondering if you have come across any potential solutions or workarounds for this issue? I would greatly appreciate any insights or suggestions you might have.

Thank you in advance for your time and help!
Here is a part of the code I provided：
`
class MambaMultiheadAttentionEncoder(nn.Module):
    def __init__(self, input_dim, num_heads, dropout_rate):
        super(MambaMultiheadAttentionEncoder, self).__init__()
        self.mamba = Mamba2Simple(d_model=input_dim, d_state=64, d_conv=4, expand=2)
        self.multiheadattention = nn.MultiheadAttention(input_dim, num_heads, dropout=dropout_rate)


        self.mlp_mamba = nn.Linear(input_dim, input_dim)
        self.mlp_attention = nn.Linear(input_dim, input_dim)

        # Dropout 层
        self.dropout = nn.Dropout(dropout_rate)
    
        self.relu = nn.LeakyReLU(negative_slope=0.01)


        self.gate_mamba = nn.Linear(input_dim, 1)
        self.gate_attention = nn.Linear(input_dim, 1)


        self.cross_attention = nn.MultiheadAttention(input_dim, num_heads, dropout=dropout_rate)


        self.mamba_weight = nn.Parameter(torch.randn(1))
        self.attn_weight = nn.Parameter(torch.randn(1))


    def forward(self, input_features):
        tmp_out = self.mlp_attention(input_features)
        # print(tmp_out)
        tmp_out = self.relu(tmp_out)
        # print(tmp_out)

        query = self.dropout(tmp_out) 
        key = self.dropout(tmp_out)  
        value = self.dropout(tmp_out)  

        multiheadattention_output, attn_output_weights = self.multiheadattention(query, key, value)
        # print(multiheadattention_output)
        multiheadattention_output = self.dropout(multiheadattention_output)  # 添加 Dropout 层

        # print(input_features)

        mamba_out = self.mamba(input_features)
        # print(mamba_out)
        mamba_out = self.mlp_mamba(mamba_out)
        mamba_out = self.relu(mamba_out)
        mamba_out = self.dropout(mamba_out)
        gate_mamba = torch.sigmoid(self.gate_mamba(mamba_out))
        gate_attention = torch.sigmoid(self.gate_attention(multiheadattention_output))

        weighted_mamba_out = gate_mamba * mamba_out
        weighted_attention_out = gate_attention * multiheadattention_output

        cross_attention_output, _ = self.cross_attention(weighted_mamba_out, weighted_attention_out, weighted_attention_out)
        tmp_out = input_features + weighted_attention_out + weighted_mamba_out + cross_attention_output
        tmp_out = self.dropout(tmp_out)  

        return tmp_out
class BertForRE(BertPreTrainedModel):
    def __init__(self, config, params):
        super().__init__(config)
        self.max_seq_len = params.max_seq_length
        self.seq_tag_size = params.seq_tag_size
        self.rel_num = params.rel_num
        self.train_batch_size = params.train_batch_size

        """pretrain model"""
        self.bert = BertModel(config)

        # pretrain model
        self.bert = BertModel(config)
        # sequence tagging
        self.sequence_tagging_sub = MultiNonLinearClassifier(config.hidden_size * 2, self.seq_tag_size, params.drop_prob)
        self.sequence_tagging_obj = MultiNonLinearClassifier(config.hidden_size * 2, self.seq_tag_size, params.drop_prob)
        # self.sequence_tagging_sum = SequenceLabelForSO(config.hidden_size, self.seq_tag_size, params.drop_prob)
        # global correspondence
        self.global_corres = MultiNonLinearClassifier(config.hidden_size * 2, 1, params.drop_prob)
        # relation judgement
        self.rel_judgement = MultiNonLinearClassifier(config.hidden_size, params.rel_num, params.drop_prob)
        self.rel_embedding = nn.Embedding(params.rel_num, config.hidden_size)

        # self.mamba = Mamba2(d_model=config.hidden_size,  d_state=64,d_conv=4,expand=2,)
        self.encoder = MambaMultiheadAttentionEncoder(config.hidden_size, 8,params.drop_prob)
        # self.mamba = Mamba2Simple(d_model=config.hidden_size, d_state=64, d_conv=4, expand=2)
        self.init_weights()


    @staticmethod
    def masked_avgpool(sent, mask):
        mask_ = mask.masked_fill(mask == 0, -1e9).float()
        score = torch.softmax(mask_, -1)
        return torch.matmul(score.unsqueeze(1), sent).squeeze(1)

    def forward(
            self,
            input_ids=None,
            attention_mask=None,
            seq_tags=None,
            potential_rels=None,
            corres_tags=None,
            rel_tags=None,
            ex_params=None
    ):
        """
        Args:
            input_ids: (batch_size, seq_len)
            attention_mask: (batch_size, seq_len)
            rel_tags: (bs, rel_num)
            potential_rels: (bs,), only in train stage.
            seq_tags: (bs, 2, seq_len)
            corres_tags: (bs, seq_len, seq_len)
            ex_params: experiment parameters
        """
        corres_threshold, rel_threshold = 0.5, 0.1
        ensure_corres, ensure_rel = True, True
        bs, seq_len = input_ids.size()
        if seq_len <=512:
            outputs = self.bert(input_ids, attention_mask=attention_mask, output_hidden_states=True)[0]
        else:
            input_id_list = torch.split(input_ids, 512, dim=1)
            # attention_mask_list = torch.split(attention_mask, 512, dim=1)
            output1 = self.bert(input_id_list[0])[0]
            output2 = self.bert(input_id_list[1])[0]
            outputs = torch.cat([output1, output2], dim=1)
        mambaout = self.encoder(outputs)
        print(mambaout)
        # print(mambaout)
        # mambaout = outputs
        bs, seq_len, h = mambaout.size()`

**Comment by liguochun0304** [comment]
Hi, I noticed that I encountered the same issue as he did. The problem I'm dealing with is that when I use the mamba module independently, it doesn't output any NaN values. However, when I integrate it into my model, NaN values start to appear. Interestingly, it works fine on one server, but when I switch to another server, the NaN values show up. Both machines are Dockerized, so I don't expect any environmental issues, and I've also double-checked things like nvcc and related configurations.

I’ve been trying to resolve this issue for a while now, but I haven’t been able to find a solution yet. I was wondering if you have come across any potential solutions or workarounds for this issue? I would greatly appreciate any insights or suggestions you might have.

Thank you in advance for your time and help!

I ran the same code on different servers and output the results at the mamba encoding step. The result was as follows:
![Snipaste_2025-01-13_10-50-08](https://github.com/user-attachments/assets/271e6d14-7dd4-4c6f-94d8-7d853f10ae89)
![Snipaste_2025-01-13_10-51-03](https://github.com/user-attachments/assets/ed793d46-d49f-4bb9-97f0-7beec326c7a5)


One outputs all NaN values, while the other outputs all zeros.

Further investigation revealed that NaN values occurred when defining the convolutional layer：
![Snipaste_2025-01-13_13-34-54](https://github.com/user-attachments/assets/c6d6dd4a-bfbd-45f7-8cf9-e88c34366b5d)
![Snipaste_2025-01-13_13-35-11](https://github.com/user-attachments/assets/b913fb0c-4293-4eec-804b-ee7c12e6f1cb)

Here is a part of the code I provided：
```
class MambaMultiheadAttentionEncoder(nn.Module):
    def __init__(self, input_dim, num_heads, dropout_rate):
        super(MambaMultiheadAttentionEncoder, self).__init__()
        self.mamba = Mamba2Simple(d_model=input_dim, d_state=64, d_conv=4, expand=2)
        self.multiheadattention = nn.MultiheadAttention(input_dim, num_heads, dropout=dropout_rate)


        self.mlp_mamba = nn.Linear(input_dim, input_dim)
        self.mlp_attention = nn.Linear(input_dim, input_dim)

        # Dropout 层
        self.dropout = nn.Dropout(dropout_rate)
    
        self.relu = nn.LeakyReLU(negative_slope=0.01)


        self.gate_mamba = nn.Linear(input_dim, 1)
        self.gate_attention = nn.Linear(input_dim, 1)


        self.cross_attention = nn.MultiheadAttention(input_dim, num_heads, dropout=dropout_rate)


        self.mamba_weight = nn.Parameter(torch.randn(1))
        self.attn_weight = nn.Parameter(torch.randn(1))


    def forward(self, input_features):
        tmp_out = self.mlp_attention(input_features)
        # print(tmp_out)
        tmp_out = self.relu(tmp_out)
        # print(tmp_out)

        query = self.dropout(tmp_out) 
        key = self.dropout(tmp_out)  
        value = self.dropout(tmp_out)  

        multiheadattention_output, attn_output_weights = self.multiheadattention(query, key, value)
        # print(multiheadattention_output)
        multiheadattention_output = self.dropout(multiheadattention_output)  # 添加 Dropout 层

        # print(input_features)

        mamba_out = self.mamba(input_features)
        # print(mamba_out)
        mamba_out = self.mlp_mamba(mamba_out)
        mamba_out = self.relu(mamba_out)
        mamba_out = self.dropout(mamba_out)
        gate_mamba = torch.sigmoid(self.gate_mamba(mamba_out))
        gate_attention = torch.sigmoid(self.gate_attention(multiheadattention_output))

        weighted_mamba_out = gate_mamba * mamba_out
        weighted_attention_out = gate_attention * multiheadattention_output

        cross_attention_output, _ = self.cross_attention(weighted_mamba_out, weighted_attention_out, weighted_attention_out)
        tmp_out = input_features + weighted_attention_out + weighted_mamba_out + cross_attention_output
        tmp_out = self.dropout(tmp_out)  

        return tmp_out
class BertForRE(BertPreTrainedModel):
    def __init__(self, config, params):
        super().__init__(config)
        self.max_seq_len = params.max_seq_length
        self.seq_tag_size = params.seq_tag_size
        self.rel_num = params.rel_num
        self.train_batch_size = params.train_batch_size

        """pretrain model"""
        self.bert = BertModel(config)

        # pretrain model
        self.bert = BertModel(config)
        # sequence tagging
        self.sequence_tagging_sub = MultiNonLinearClassifier(config.hidden_size * 2, self.seq_tag_size, params.drop_prob)
        self.sequence_tagging_obj = MultiNonLinearClassifier(config.hidden_size * 2, self.seq_tag_size, params.drop_prob)
        # self.sequence_tagging_sum = SequenceLabelForSO(config.hidden_size, self.seq_tag_size, params.drop_prob)
        # global correspondence
        self.global_corres = MultiNonLinearClassifier(config.hidden_size * 2, 1, params.drop_prob)
        # relation judgement
        self.rel_judgement = MultiNonLinearClassifier(config.hidden_size, params.rel_num, params.drop_prob)
        self.rel_embedding = nn.Embedding(params.rel_num, config.hidden_size)

        # self.mamba = Mamba2(d_model=config.hidden_size,  d_state=64,d_conv=4,expand=2,)
        self.encoder = MambaMultiheadAttentionEncoder(config.hidden_size, 8,params.drop_prob)
        # self.mamba = Mamba2Simple(d_model=config.hidden_size, d_state=64, d_conv=4, expand=2)
        self.init_weights()


    @staticmethod
    def masked_avgpool(sent, mask):
        mask_ = mask.masked_fill(mask == 0, -1e9).float()
        score = torch.softmax(mask_, -1)
        return torch.matmul(score.unsqueeze(1), sent).squeeze(1)

    def forward(
            self,
            input_ids=None,
            attention_mask=None,
            seq_tags=None,
            potential_rels=None,
            corres_tags=None,
            rel_tags=None,
            ex_params=None
    ):
        """
        Args:
            input_ids: (batch_size, seq_len)
            attention_mask: (batch_size, seq_len)
            rel_tags: (bs, rel_num)
            potential_rels: (bs,), only in train stage.
            seq_tags: (bs, 2, seq_len)
            corres_tags: (bs, seq_len, seq_len)
            ex_params: experiment parameters
        """
        corres_threshold, rel_threshold = 0.5, 0.1
        ensure_corres, ensure_rel = True, True
        bs, seq_len = input_ids.size()
        if seq_len <=512:
            outputs = self.bert(input_ids, attention_mask=attention_mask, output_hidden_states=True)[0]
        else:
            input_id_list = torch.split(input_ids, 512, dim=1)
            # attention_mask_list = torch.split(attention_mask, 512, dim=1)
            output1 = self.bert(input_id_list[0])[0]
            output2 = self.bert(input_id_list[1])[0]
            outputs = torch.cat([output1, output2], dim=1)
        mambaout = self.encoder(outputs)
        print(mambaout)
        # print(mambaout)
        # mambaout = outputs
        bs, seq_len, h = mambaout.size()```

**Comment by TheMakerOfWorlds** [comment]
same problem cant find any fix

**Comment by Sheng-T** [comment]
I also have the same problem
